data-structures/ds-avltree/c/AVLRouteInC.c

```diff
@@ -0,0 +1,255 @@
+*/
+
+#include <stdio.h>
+#include <stdlib.h>
+
+
+typedef struct node{
+	int num;
+	struct node* left;
+	struct node* right;
+}Node;
+
+typedef Node* TreeBin;
+
+TreeBin * newTree(){
+	TreeBin* root = (TreeBin*) malloc(sizeof(TreeBin));
+	if(root)
+	    (*root) = NULL;
+
+	return root;
+}
+
+void clearNode(Node* node){
+	if(!node)	return;
+	clearNode(node->left);
+	clearNode(node->right);
+	free(node);
+	node = NULL;
+}
+
+void clearTree(TreeBin* Tree){
+	if(!Tree)	return;
+	clearNode(*Tree);
+	free(Tree);
+}
+
+int isEmpty(TreeBin* Tree){
+	if(!Tree)	return 1;
+	if(!(*Tree))	return 1;
+	return 0;
+}
+
+int countNode(TreeBin* Tree){
+	int tleft, tright;
+
+	if(isEmpty(Tree))	return 0;
+
+	tleft = countNode(&((*Tree)->left));
+	tright = countNode(&((*Tree)->right));
+
+	return (tleft+tright+1);
+}
+
+int heightTree(TreeBin* Tree){
+	int heightLeft, heightRight;
+
+	if(isEmpty(Tree))	return -1;
+	if(isEmpty(&((*Tree)->left)) && isEmpty(&((*Tree)->right)))	return 0;
+
+	heightLeft = heightTree(&((*Tree)->left));
+	heightRight = heightTree(&((*Tree)->right));
+
+	if(heightLeft > heightRight)	return (heightLeft + 1);
+
+	return (heightRight + 1);
+}
+
+//balancing functions
+// [-1, 1] balanced tree
+// [-2, 2] not balanced tree
+int Balanced(TreeBin *Tree){
+	return heightTree(&((*Tree)->left)) - heightTree(&((*Tree)->right));
+}
+
+void spinRight(TreeBin *Tree){
+	Node *nod;
+	nod = (*Tree)->left;
+	(*Tree)->left = nod->right;
+	nod->right = (*Tree);
+	(*Tree) = nod;
+}
+
+void spinLeft(TreeBin *Tree){
+	Node *nod;
+	nod = (*Tree)->right;
+	(*Tree)->right = nod->left;
+	nod->left = (*Tree);
+	(*Tree) = nod;
+}
+
+void balance(TreeBin *Tree, int n){
+	int eb, it;
+
+	if(!(*Tree))    return;
+
+	eb = Balanced(Tree);
+	if(eb >= -1 && eb <= 1)	return;
+
+	else{
+
+		if(n > (*Tree)->num){
+			it = Balanced(&((*Tree)->right));
+
+			if(eb <= -2){
+				if(it == -1 || !it)	spinLeft(Tree);
+
+				else{	
+				    spinRight(&((*Tree)->right)); 
+				    spinLeft(Tree);
+				}
+			}else{
+
+				if(it == 1 || !it)	spinRight(Tree);
+
+				else{	
+				    spinLeft(&((*Tree)->right)); 
+				    spinRight(Tree);
+
+				}
+			}
+		}else{
+
+			it = Balanced(&((*Tree)->left));
+
+			if(eb <= -2){
+				if(it == -1 || !it)	spinLeft(Tree);
+				else{	
+				    spinRight(&((*Tree)->left)); 
+				    spinLeft(Tree);
+				}
+			}else{
+				if(it == 1 || !it)	spinRight(Tree);
+				else{	
+				    spinLeft(&((*Tree)->left)); 
+				    spinRight(Tree);
+
+				}
+			}
+		}
+		return;
+	}
+}
+
+void insertTree(TreeBin* Tree, int n){
+	Node *nod;
+	if(!Tree)	return ;
+	if(!(*Tree)){
+		nod = (Node*)malloc(sizeof(Node));
+		nod->num = n;
+		nod->left = NULL;
+		nod->right = NULL;
+		(*Tree) = nod;
+		return;
+	}
+	if(n > (*Tree)->num)	
+	    insertTree(&((*Tree)->right), n);
+	else	
+	    insertTree(&((*Tree)->left), n);
+
+	balance(Tree, n);
+}
+
+//Get the higher element of the left sub-tree
+void subleftTree(TreeBin *Tree, TreeBin *next){
+	Node *nod;
```

Comment:
Codacy found an issue: [The scope of the variable 'nod' can be reduced.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6412547)